README.md
# Edge

[![Build Status](https://travis-ci.org/jackc/edge.svg?branch=master)](https://travis-ci.org/jackc/edge)

Edge provides graph functionality to ActiveRecord using recursive common table
expressions. It has only been tested with PostgreSQL, but it uses Arel for
SQL generation so it should work with any database and adapter that support
recursive CTEs.

acts_as_forest enables an entire tree or even an entire forest of trees to
be loaded in a single query. All parent and children associations are
preloaded.

## Installation

Add this line to your application's Gemfile:

    gem 'edge'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install edge

## Usage

acts_as_forest adds tree / multi-tree functionality. All it needs a parent_id
column. This can be overridden by passing a :foreign_key option to
acts_as_forest.

    class Location < ActiveRecord::Base
      acts_as_forest :order => "name"
    end

    usa = Location.create! :name => "USA"
    illinois = usa.children.create! :name => "Illinois"
    chicago = illinois.children.create! :name => "Chicago"
    indiana = usa.children.create! :name => "Indiana"
    canada = Location.create! :name => "Canada"
    british_columbia = canada.children.create! :name => "British Columbia"

    Location.root.all # [usa, canada]
    Location.find_forest # [usa, canada] with all children and parents preloaded
    Location.find_tree usa.id # load a single tree.

It also provides the with_descendants scope to get all currently selected
nodes and all their descendents. It can be chained after where scopes, but
must not be used after any other type of scope.

    Location.where(name: "Illinois").with_descendants.all # [illinois, chicago]

Also supported is `ancestors` instance method which returns an array of all ancestors ordered by nearest ancestors first.

    city = Location.where(name: 'Chicago')
    ancestors = city.ancestors # [illinois, usa]

## Benchmarks

Edge includes a performance benchmarks. You can create test forests with a
configurable number of trees, depth, number of children per node, and
size of payload per node.

    jack@moya:~/work/edge$ ruby -I lib -I bench bench/forest_find.rb --help
    Usage: forest_find [options]
        -t, --trees NUM                  Number of trees to create
        -d, --depth NUM                  Depth of trees
        -c, --children NUM               Number of children per node
        -p, --payload NUM                Characters of payload per node

Even on slower machines entire trees can be loaded quickly.

    jack@moya:~/work/edge$ ruby -I lib -I bench bench/forest_find.rb
    Trees: 50
    Depth: 3
    Children per node: 10
    Payload characters per node: 16
    Descendants per tree: 110
    Total records: 5550
                                                   user     system      total        real
    Load entire forest 10 times                4.260000   0.010000   4.270000 (  4.422442)
    Load one tree 100 times                    0.830000   0.040000   0.870000 (  0.984642)

### Running the benchmarks

1. Create a database such as edge_bench.
2. Configure bench/database.yml to connect to it.
3. Load bench/database_structure.sql into your bench database.
4. Run benchmark scripts from root of gem directory (remember to pass ruby
   the include paths for lib and bench)

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License

MIT


CHANGELOG.md
# 0.6.1 (February 20, 2021)

* Add Ruby 3 compatibility (Martins Polakovs)

# 0.6.0 (December 13, 2018)

* Fix usage when table name includes schema
* Drop Rails 4.x support

# 0.5.1 (August 4, 2017)

* Allow use of 'optional' option for belongs_to (Yuji Yaginuma)

# 0.5.0 (May 14, 2016)

* Add Rails 5.0 support
* Drop Rails 4.0 support
* Replace using internal arel with SQL string building

# 0.4.4 (January 29, 2016)

* Fix JRuby compatibility (jackc)

# 0.4.3 (October 23, 2015)

* Allow dependent option to acts_as_forest (WANG QUANG)

# 0.4.2 (May 21, 2015)

* Fixed premature SQL-ization that could result in PG protocol violation errors (Neil E. Pearson)
* Require Rails 4.0+
* Document ancestors method (science)

# 0.4.1 (January 15, 2015)

* Include rake as development dependency
* Fix for not passing string to belongs_to class_name (davekaro)

# 0.4.0 (December 23, 2014)

* Fix failure with bind_values
* Fix README typos (y-yagi)
* Improve performance by using flat_map (TheKidCoder)

# 0.3.2 (March 1, 2014)

* Set inverse_of on parent and children associations (Systho)

# 0.3.1 (February 7, 2014)

* Allow includes and with_descendents to work together (Systho)

# 0.3.0 (December 17, 2013)

* Rails 4 support
* Fix for incomparable column types

# 0.2.1 (March 6, 2013)

* Fix: acts_as_forest survives multiple calls

# 0.2.0 (February 22, 2013)

* Added with_descendents

# 0.1.0 (March 11, 2012)

* Initial release


spec/forest_spec.rb
require 'spec_helper'

class Location < ActiveRecord::Base
  acts_as_forest :order => "name"
end

class BodyPart < ActiveRecord::Base
  acts_as_forest :foreign_key => "body_part_id"
end

Location.delete_all

describe "Edge::Forest" do
  let(:skeleton) { BodyPart.create! }
  let!(:usa) { Location.create! :name => "USA" }
  let!(:illinois) { Location.create! :parent => usa, :name => "Illinois" }
  let!(:chicago) { Location.create! :parent => illinois, :name => "Chicago" }
  let!(:indiana) { Location.create! :parent => usa, :name => "Indiana" }
  let!(:canada) { Location.create! :name => "Canada" }
  let!(:british_columbia) { Location.create! :parent => canada, :name => "British Columbia" }

  describe "root?" do
    context "of root node" do
      it "should be true" do
        expect(usa.root?).to eq true
      end
    end

    context "of model with custom foreign key" do
      it "should be true" do
        expect(skeleton.root?).to eq true
      end
    end

    context "of child node" do
      it "should be false" do
        expect(illinois.root?).to eq false
      end
    end

    context "of leaf node" do
      it "should be root node" do
        expect(chicago.root?).to eq false
      end
    end
  end

  describe "root" do
    context "of root node" do
      it "should be self" do
        expect(usa.root).to eq usa
      end
    end

    context "of child node" do
      it "should be root node" do
        expect(illinois.root).to eq usa
      end
    end

    context "of leaf node" do
      it "should be root node" do
        expect(chicago.root).to eq usa
      end
    end
  end

  describe "parent" do
    context "of root node" do
      it "should be nil" do
        expect(usa.parent).to eq nil
      end
    end

    context "of child node" do
      it "should be parent" do
        expect(illinois.parent).to eq usa
      end
    end

    context "of leaf node" do
      it "should be parent" do
        expect(chicago.parent).to eq illinois
      end
    end
  end

  describe "ancestors" do
    context "of root node" do
      it "should be empty" do
        expect(usa.ancestors).to be_empty
      end
    end

    context "of leaf node" do
      it "should be ancestors ordered by ascending distance" do
        expect(chicago.ancestors).to eq [illinois, usa]
      end
    end
  end

  describe "siblings" do
    context "of root node" do
      it "should be empty" do
        expect(usa.siblings).to be_empty
      end
    end

    context "of child node" do
      it "should be other children of parent" do
        expect(illinois.siblings).to include(indiana)
      end
    end
  end

  describe "children" do
    it "should be children" do
      expect(usa.children).to include(illinois, indiana)
    end

    it "should be ordered" do
      alabama = Location.create! :parent => usa, :name => "Alabama"
      expect(usa.children).to eq [alabama, illinois, indiana]
    end

    context "of leaf" do
      it "should be empty" do
        expect(chicago.children).to be_empty
      end
    end
  end

  describe "descendants" do
    it "should be all descendants" do
      expect(usa.descendants).to include(illinois, indiana, chicago)
    end

    context "of leaf" do
      it "should be empty" do
        expect(chicago.descendants).to be_empty
      end
    end
  end

  describe "root scope" do
    it "returns only root nodes" do
      expect(Location.root).to include(usa, canada)
    end
  end

  describe "find_forest" do
    it "preloads all parents and children" do
      forest = Location.find_forest

      Location.where("purposely fail if any Location find happens here").scoping do
        forest.each do |tree|
          tree.descendants.each do |node|
            expect(node.parent).to be
            expect(node.children).to be_kind_of(ActiveRecord::Associations::CollectionProxy)
          end
        end
      end
    end

    it "works when scoped" do
      forest = Location.where(:name => "USA").find_forest
      expect(forest).to match_array([usa])
      expect(forest.first.children).to match_array([illinois, indiana])
    end

    it "preloads children in proper order" do
      alabama = Location.create! :parent => usa, :name => "Alabama"
      forest = Location.find_forest
      tree = forest.find { |l| l.id == usa.id }
      expect(tree.children).to eq [alabama, illinois, indiana]
    end

    context "with an infinite loop" do
      before do
        usa.update_attribute(:parent, chicago)
      end

      it "does not re-loop" do
        Location.find_forest
      end
    end
  end

  describe "find_tree" do
    it "finds by id" do
      tree = Location.find_tree usa.id
      expect(tree).to eq usa
    end

    it "finds multiple trees by id" do
      trees = Location.find_tree [indiana.id, illinois.id]
      expect(trees).to include(indiana, illinois)
    end

    it "raises ActiveRecord::RecordNotFound when id is not found" do
      expect{Location.find_tree -1}.to raise_error(ActiveRecord::RecordNotFound)
    end

    it "raises ActiveRecord::RecordNotFound when not all ids are not found" do
      expect{Location.find_tree [indiana.id, -1]}.to raise_error(ActiveRecord::RecordNotFound)
    end

  end

  describe "with_descendants" do
    context "unscoped" do
      it "returns all records" do
        rows = Location.with_descendants.to_a
        expect(rows).to match_array Location.all
      end
    end

    context "scoped" do
      it "returns a new scope that includes previously scoped records and their descendants" do
        rows = Location.where(id: canada.id).with_descendants.to_a
        expect(rows).to match_array [canada, british_columbia]
      end

      it "is not commutative" do
        rows = Location.with_descendants.where(id: canada.id).to_a
        expect(rows).to eq [canada]
      end
    end
  end

  describe "self.acts_as_forest" do
    it 'can be used twice' do
      class Location2 < ActiveRecord::Base
        self.table_name = 'locations'
        acts_as_forest :order => "name"
      end

      Location2.find_forest
    end
  end

  describe "dependent destroy" do
    it 'cascades destroys' do
      class Location3 < ActiveRecord::Base
        self.table_name = 'locations'
        acts_as_forest dependent: :destroy
      end

      Location3.find(usa.id).destroy

      expect(Location.exists?(usa.id)).to eq false
      expect(Location.exists?(illinois.id)).to eq false
      expect(Location.exists?(chicago.id)).to eq false
      expect(Location.exists?(indiana.id)).to eq false

      expect(Location.exists?(canada.id)).to eq true
      expect(Location.exists?(british_columbia.id)).to eq true
    end

  end

  if ActiveRecord::VERSION::MAJOR >= 5
    describe "optional option" do
      before do
        @original_value = ActiveRecord::Base.belongs_to_required_by_default
        ActiveRecord::Base.belongs_to_required_by_default = true
      end

      after do
        ActiveRecord::Base.belongs_to_required_by_default = @original_value
      end

      it 'parent can be nil' do
        class Location4 < ActiveRecord::Base
          self.table_name = "locations"
          acts_as_forest optional: true
        end

        expect(Location4.new(name: "Iceland").valid?).to eq true
      end
    end
  end
end


spec/spec_helper.rb
require 'edge'
require 'yaml'

require 'rspec'

database_config = YAML.load_file(File.expand_path("../database.yml", __FILE__))
ActiveRecord::Base.establish_connection database_config["test"]

RSpec.configure do |config|
  config.before(:all) do |example|
    ActiveRecord::Base.connection.execute <<-SQL
      truncate body_parts;
      truncate locations;
    SQL
  end

  config.around do |example|
    ActiveRecord::Base.transaction do
      example.call
      raise ActiveRecord::Rollback
    end
  end
end


lib/edge/forest.rb
module Edge
  module Forest
    # acts_as_forest models a tree/multi-tree structure.
    module ActsAsForest
      # options:
      #
      # * dependent - passed to children has_many (default: none)
      # * foreign_key - column name to use for parent foreign_key (default: parent_id)
      # * order - how to order children (default: none)
      # * optional - passed to belongs_to (default: none)
      def acts_as_forest(options={})
        options.assert_valid_keys :foreign_key, :order, :dependent, :optional

        class_attribute :forest_foreign_key
        self.forest_foreign_key = options[:foreign_key] || "parent_id"

        class_attribute :forest_order
        self.forest_order = options[:order] || nil

        common_options = {
          :class_name => self.name,
          :foreign_key => forest_foreign_key
        }

        dependent_options = options[:dependent] ? { dependent: options[:dependent] } : {}

        optional_options = options[:optional] ? { optional: options[:optional] } : {}

        belongs_to :parent, **common_options.merge(inverse_of: :children).merge(optional_options)

        if forest_order
          has_many :children, -> { order(forest_order) }, **common_options.merge(inverse_of: :parent).merge(dependent_options)
        else
          has_many :children, **common_options.merge(inverse_of: :parent).merge(dependent_options)
        end

        scope :root, -> { where(forest_foreign_key => nil) }

        include Edge::Forest::InstanceMethods
        extend Edge::Forest::ClassMethods
      end
    end

    module ClassMethods
      # Finds entire forest and preloads all associations. It can be used at
      # the end of an ActiveRecord finder chain.
      #
      # Example:
      #    # loads all locations
      #    Location.find_forest
      #
      #    # loads all nodes with matching names and all there descendants
      #    Category.where(:name => %w{clothing books electronics}).find_forest
      def find_forest
        new_scope = unscoped.joins("INNER JOIN all_nodes USING(#{connection.quote_column_name primary_key})")
        new_scope = new_scope.order(forest_order) if forest_order

        sql = <<-SQL
          #{cte_sql}
          #{new_scope.to_sql}
        SQL
        records = find_by_sql sql

        records_by_id = records.each_with_object({}) { |r, h| h[r.id] = r }

        # Set all children associations to an empty array
        records.each do |r|
          children_association = r.association(:children)
          children_association.target = []
        end

        top_level_records = []

        records.each do |r|
          parent = records_by_id[r[forest_foreign_key]]
          if parent
            r.association(:parent).target = parent
            parent.association(:children).target.push(r)
          else
            top_level_records.push(r)
          end
        end

        top_level_records
      end

      # Finds an a tree or trees by id.
      #
      # If any requested ids are not found it raises
      # ActiveRecord::RecordNotFound.
      def find_tree(id_or_ids)
        trees = where(:id => id_or_ids).find_forest
        if id_or_ids.kind_of?(Array)
          raise ActiveRecord::RecordNotFound unless trees.size == id_or_ids.size
          trees
        else
          raise ActiveRecord::RecordNotFound if trees.empty?
          trees.first
        end
      end

      # Returns a new scope that includes previously scoped records and their descendants by subsuming the previous scope into a subquery
      #
      # Only where scopes can precede this in a scope chain
      def with_descendants
        subquery_scope = unscoped
          .joins("INNER JOIN all_nodes USING(#{connection.quote_column_name primary_key})")
          .select(primary_key)

        subquery_sql = <<-SQL
          #{cte_sql}
          #{subquery_scope.to_sql}
        SQL

        unscoped.where <<-SQL
          #{connection.quote_column_name primary_key} IN (#{subquery_sql})
        SQL
      end

      private
      def cte_sql
        quoted_table_name = '"locations"'
        original_scope = (current_scope || all).select(primary_key, forest_foreign_key)
        iterated_scope = unscoped.select(primary_key, forest_foreign_key)
          .joins("INNER JOIN all_nodes ON #{connection.quote_table_name table_name}.#{connection.quote_column_name forest_foreign_key}=all_nodes.#{connection.quote_column_name primary_key}")
        <<-SQL
          WITH RECURSIVE all_nodes AS (
            #{original_scope.to_sql}
            UNION
            #{iterated_scope.to_sql}
          )
        SQL
      end
    end

    module InstanceMethods
      # Returns the root of this node. If this node is root returns self.
      def root
        parent ? parent.root : self
      end

      # Returns true is this node is a root or false otherwise
      def root?
        !self[forest_foreign_key]
      end

      # Returns all sibling nodes (nodes that have the same parent). If this
      # node is a root node it returns an empty array.
      def siblings
        parent ? parent.children - [self] : []
      end

      # Returns all ancestors ordered by nearest ancestors first.
      def ancestors
        _ancestors = []
        node = self
        while(node = node.parent)
          _ancestors.push(node)
        end

        _ancestors
      end

      # Returns all descendants
      def descendants
        if children.present?
          children + children.flat_map(&:descendants)
        else
          []
        end
      end
    end
  end
end

ActiveRecord::Base.extend Edge::Forest::ActsAsForest


bench/forest_find.rb
require 'benchmark_helper'

options = {}
optparse = OptionParser.new do |opts|
  options[:num_trees] = 50
  opts.on '-t NUM', '--trees NUM', Integer, 'Number of trees to create' do |n|
    options[:num_trees] = n
  end
  
  options[:depth] = 3
  opts.on '-d NUM', '--depth NUM', Integer, 'Depth of trees' do |n|
    options[:depth] = n
  end
  
  options[:num_children] = 10
  opts.on '-c NUM', '--children NUM', Integer, 'Number of children per node' do |n|
    options[:num_children] = n
  end
  
  options[:payload_size] = 16
  opts.on '-p NUM', '--payload NUM', Integer, 'Characters of payload per node' do |n|
    options[:payload_size] = n
  end
end

optparse.parse!

NUM_TREES = options[:num_trees]
DEPTH = options[:depth]
NUM_CHILDREN = options[:num_children]
PAYLOAD_SIZE = options[:payload_size]


def create_forest_tree(current_depth = 1, parent = nil)
  node = ActsAsForestRecord.create! :parent => parent, :payload => "z" * PAYLOAD_SIZE
  unless current_depth == DEPTH
    NUM_CHILDREN.times { create_forest_tree current_depth + 1, node }
  end
  node
end

clean_database
ActsAsForestRecord.transaction do
  NUM_TREES.times { create_forest_tree }
end
vacuum_analyze

puts "Trees: #{NUM_TREES}"
puts "Depth: #{DEPTH}" 
puts "Children per node: #{NUM_CHILDREN}"
puts "Payload characters per node: #{PAYLOAD_SIZE}"
puts "Descendants per tree: #{ActsAsForestRecord.find_tree(ActsAsForestRecord.root.first.id).descendants.size}"
puts "Total records: #{ActsAsForestRecord.count}"


Benchmark.bm(40) do |x|
  load_entire_forest_times = 10
  x.report("Load entire forest #{load_entire_forest_times} times") do
    load_entire_forest_times.times do
      ActsAsForestRecord.find_forest
    end
  end
  
  load_one_tree_times = 100
  first_tree_id = ActsAsForestRecord.root.first.id
  x.report("Load one tree #{load_one_tree_times} times") do
    load_one_tree_times.times do
      ActsAsForestRecord.find_tree first_tree_id
    end
  end
end
